release/src/milestones.ts

```diff
@@ -0,0 +1,166 @@
+import { getMilestones } from "./github";
+import { getLinkedIssues, getPRsFromCommitMessage } from "./linked-issues";
+import type { Issue, GithubProps, Milestone } from "./types";
+import {
+  getMajorVersion,
+  getVersionFromReleaseBranch,
+} from "./version-helpers";
+
+function isBackport(pullRequest: Issue) {
+  return pullRequest.title.includes('backport') ||
+    (
+      Array.isArray(pullRequest.labels) &&
+      pullRequest.labels.some((label) => label.name === 'was-backported')
+    );
+}
+
+// for auto-setting milestones, we don't ever want to auto-set a patch milestone
+// which we release VERY rarely
+function ignorePatches(version: string) {
+  return version.split('.').length < 4;
+}
+
+function versionSort(a: string, b: string) {
+  const [aMajor, aMinor] = a.split('.').map(Number);
+  const [bMajor, bMinor] = b.split('.').map(Number);
+
+  if (aMajor !== bMajor) {
+    return aMajor - bMajor;
+  }
+
+  if (aMinor !== bMinor) {
+    return aMinor - bMinor;
+  }
+
+  return 0;
+}
+
+const isNotNull = <T>(value: T | null): value is T => value !== null;
+
+
+async function getOriginalPR({
+  github,
+  repo,
+  owner,
+  pullRequestNumber,
+}: GithubProps & { pullRequestNumber: number }) {
+  // every PR in the release branch should have a pr number
+  // it could be a backport PR or an original PR
+  const pull = await github.rest.pulls.get({
+    owner,
+    repo,
+    pull_number: pullRequestNumber,
+  });
+
+  if (pull?.data && isBackport(pull.data)) {
+    return getOriginalPR({
+      github,
+      repo,
+      owner,
+      pullRequestNumber: pull.data.number
+    });
+  }
+
+  const linkedIssues = await getLinkedIssues(pull.data.body ?? '');
+
+  if (linkedIssues) {
+    console.log('found linked issue for PR', pull.data.number, linkedIssues);
+    return linkedIssues.map(Number);
+  }
+  console.log("no linked issues found in PR body", pull.data.number);
+  return [pull.data.number];
+}
+
+async function setMilestone({ github, owner, repo, issueNumber, milestone }: GithubProps & { issueNumber: number, milestone: Milestone }) {
+  // we can use this for both issues and PRs since they're the same for many purposes in github
+  const issue = await github.rest.issues.get({
+    owner,
+    repo,
+    issue_number: issueNumber,
+  });
+
+  if (!issue.data.milestone) {
+    return github.rest.issues.update({
+      owner,
+      repo,
+      issue_number: issueNumber,
+      milestone: milestone.number,
+    });
+  }
+
+  if (issue.data.milestone.number === milestone.number) {
+    console.log(`Issue ${issueNumber} is already tagged with this ${milestone.title} milestone`);
+    return;
+  }
+
+  if (issue.data.milestone.number !== milestone.number) {
+    console.log(`Issue ${issueNumber} is already tagged with a different milestone: ${issue.data.milestone.title}`);
+
+    return github.rest.issues.createComment({
```

Comment:
Suggested tweak: if the new milestone is for an earlier version than the current one, then I would update the milestone property and post the previous milestone in the comment. This will typically be the case.

I think people have been socialized to look at the milestone to see the oldest release with an issue being fixed, and even if we learn to scour the comments instead, that's less efficient.

I guess one downside here is if the milestone had been manually set as a blocker for a given version, it's surprising and risky for it to change, but typically we backport from newest to oldest so this is unlikely to trip us up.

Reply:
I'm not sure what the right behavior actually is here. I'll ask Success what they would want in this situation.

Reply:
heard from success. they always want the oldest milestone on the issue 👍

Reply:
<img width="934" alt="Screen Shot 2024-06-19 at 4 46 46 PM" src="https://github.com/metabase/metabase/assets/30528226/4681aa60-9d04-4f58-a35b-f61fa1ff1058">